### PR TITLE
ts-web/core: grpc-web error wrapping

### PR DESCRIPTION
**Pull request by pro-wh**
cc https://github.com/oasisprotocol/oasis-wallet-web/pull/498

### CODE REVIEW

**Comment by pro-wh** [comment]
unfinished. goals:

- keep oasisCode and oasisModule available at top level error
- always give proper Error to outside, not plain object
- continue offering rewritten message if application-level message is blank
- add information about the method and url used
- add a copy of the stack from before the call, for certain bad browsers that don't support async stack and have enormous market share

**Comment by pro-wh** [comment]
preview of this in the web wallet:

extra_stack

the extra stack part will look better after https://github.com/oasisprotocol/oasis-wallet-web/pull/517

but unfortunately the way redux-saga has us yield a `call` action and not actually do it directly inside the saga, I must ask if we really want this "extra stack"

**Comment by pro-wh** [comment]
cc @lukaw3d

File: client-sdk/ts-web/core/src/client.ts [hunk 1]
```diff
@@ -569,7 +574,15 @@ export class GRPCWrapper {
     ): Promise<RESP> {
         // @ts-expect-error missing declaration
         const name = desc.name;
-        return this.client.thenableCall(this.base + name, request, null, desc).catch((e) => {
+        const method = this.base + name;
+        // Some browsers with enormous market share aren't able to preserve the stack between here
+        // and our `.catch` callback below. Save a copy explicitly.
+        // In particular, our build process rewrites async code as some generator thing, and that
+        // breaks those aforementioned browsers' tracking of async stacks. If in the future we
+        // start shipping native async/await code, let's remove this and conver callUnary to
```
Comment (lukaw3d) [inline client-sdk/ts-web/core/src/client.ts:582]:
```suggestion
        // start shipping native async/await code, let's remove this and convert callUnary to
```

Comment (pro-wh) [reply client-sdk/ts-web/core/src/client.ts]: huh, although tangentially even without converting these to async we already get to see all the way out to `runCallEffect`

Comment (pro-wh) [reply client-sdk/ts-web/core/src/client.ts]: oh shoot, its probably not that, with the bottom of the extra stack mixing together with the real stack

**Comment by pro-wh** [comment]
adding a `---` to show where extra stack is separate from the wrapped error

```
Extra stack:
Error
    at NodeInternal.callUnary (http://localhost:5000/static/js/main.701bad0c.chunk.js:2845:24)
    at NodeInternal.stakingAccount (http://localhost:5000/static/js/main.701bad0c.chunk.js:3300:19)
    at runCallEffect (http://localhost:5000/static/js/2.946a7329.chunk.js:74103:21)
    at runEffect (http://localhost:5000/static/js/2.946a7329.chunk.js:74780:7)
    at digestEffect (http://localhost:5000/static/js/2.946a7329.chunk.js:74847:5)
    at next (http://localhost:5000/static/js/2.946a7329.chunk.js:74737:9)
    at proc (http://localhost:5000/static/js/2.946a7329.chunk.js:74684:3)
    at runEffect (http://localhost:5000/static/js/2.946a7329.chunk.js:74775:7)
    at digestEffect (http://localhost:5000/static/js/2.946a7329.chunk.js:74847:5)
    at http://localhost:5000/static/js/2.946a7329.chunk.js:74251:5
---
    at http://localhost:5000/static/js/main.701bad0c.chunk.js:2884:18
```

**Comment by pro-wh** [comment]
with async now, production build:

```
Extra stack:
Error
    at NodeInternal._callee$ (http://localhost:5000/static/js/main.9cbd1b88.chunk.js:2856:30)
    at tryCatch (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:41088:17)
    at Generator.invoke [as _invoke] (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:41305:22)
    at Generator.next (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:41141:21)
    at asyncGeneratorStep (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:19683:24)
    at _next (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:19705:9)
    at http://localhost:5000/static/js/2.5e34cd3a.chunk.js:19712:7
    at new Promise (<anonymous>)
    at NodeInternal.<anonymous> (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:19701:12)
    at NodeInternal.callUnary (http://localhost:5000/static/js/main.9cbd1b88.chunk.js:2924:27)
---
    at NodeInternal._callee$ (http://localhost:5000/static/js/main.9cbd1b88.chunk.js:2910:22)
    at tryCatch (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:41088:17)
    at Generator.invoke [as _invoke] (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:41305:22)
    at Generator.throw (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:41141:21)
    at asyncGeneratorStep (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:19683:24)
    at _throw (http://localhost:5000/static/js/2.5e34cd3a.chunk.js:19709:9)
```

**Comment by pro-wh** [comment]
with async, dev build (no transpile to generator):

```
Extra stack:
Error
    at NodeInternal.callUnary (http://localhost:3000/static/js/main.chunk.js:477:24)
    at NodeInternal.stakingAccount (http://localhost:3000/static/js/main.chunk.js:892:17)
    at runCallEffect (http://localhost:3000/static/js/vendors~main.chunk.js:22864:21)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at proc (http://localhost:3000/static/js/vendors~main.chunk.js:23449:3)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23540:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at http://localhost:3000/static/js/vendors~main.chunk.js:23012:5
    at Array.forEach (<anonymous>)
    at runAllEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23011:8)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at currCb (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
    at http://localhost:3000/static/js/vendors~main.chunk.js:22931:9
    at immediately (http://localhost:3000/static/js/vendors~main.chunk.js:22385:12)
    at runForkEffect (http://localhost:3000/static/js/vendors~main.chunk.js:22923:3)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at currCb (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
    at http://localhost:3000/static/js/vendors~main.chunk.js:22931:9
    at immediately (http://localhost:3000/static/js/vendors~main.chunk.js:22385:12)
    at runForkEffect (http://localhost:3000/static/js/vendors~main.chunk.js:22923:3)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at Object.currCb [as cont] (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
    at end (http://localhost:3000/static/js/vendors~main.chunk.js:23353:10)
    at Object.task.cont (http://localhost:3000/static/js/vendors~main.chunk.js:23174:11)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23511:18)
    at currCb (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
    at runSelectEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23070:5)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at proc (http://localhost:3000/static/js/vendors~main.chunk.js:23449:3)
    at runCallEffect (http://localhost:3000/static/js/vendors~main.chunk.js:22873:7)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at currCb (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
---
    at NodeInternal.callUnary (http://localhost:3000/static/js/main.chunk.js:524:18)
```

**Comment by pro-wh** [comment]
Huh, turns out it's no better to switch to async. even if we keep the native async+await, the error still only gets one frame

File: client-sdk/ts-web/core/src/client.ts (hunk 1 above)
Comment (pro-wh) [reply client-sdk/ts-web/core/src/client.ts]: this turned out to be wrong anyway. I'll remove the comment

**Comment by pro-wh** [comment]
ok I changed it to "invocation stack" and added an "end of invocation stack" marker at the end of the message
invocation_stack

**Comment by pro-wh** [comment]
hm, kind of weird: the caller of callUnary, client.stakingAccount wasn't `async` above. in the debug build, with callUnary async and with stakingAccount async:

```
Error: callUnary method https://grpc.oasis.dev/oasis-core.Staking/Account: RpcError: Http response at 400 or 500 level
Invocation stack:
Error
    at NodeInternal.callUnary (http://localhost:3000/static/js/main.chunk.js:473:29)
    at NodeInternal.stakingAccount (http://localhost:3000/static/js/main.chunk.js:889:23)
    at runCallEffect (http://localhost:3000/static/js/vendors~main.chunk.js:22864:21)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at proc (http://localhost:3000/static/js/vendors~main.chunk.js:23449:3)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23540:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at http://localhost:3000/static/js/vendors~main.chunk.js:23012:5
    at Array.forEach (<anonymous>)
    at runAllEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23011:8)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at currCb (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
    at http://localhost:3000/static/js/vendors~main.chunk.js:22931:9
    at immediately (http://localhost:3000/static/js/vendors~main.chunk.js:22385:12)
    at runForkEffect (http://localhost:3000/static/js/vendors~main.chunk.js:22923:3)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at currCb (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
    at http://localhost:3000/static/js/vendors~main.chunk.js:22931:9
    at immediately (http://localhost:3000/static/js/vendors~main.chunk.js:22385:12)
    at runForkEffect (http://localhost:3000/static/js/vendors~main.chunk.js:22923:3)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at Object.currCb [as cont] (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
    at end (http://localhost:3000/static/js/vendors~main.chunk.js:23353:10)
    at Object.task.cont (http://localhost:3000/static/js/vendors~main.chunk.js:23174:11)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23511:18)
    at currCb (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
    at runSelectEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23070:5)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at proc (http://localhost:3000/static/js/vendors~main.chunk.js:23449:3)
    at runCallEffect (http://localhost:3000/static/js/vendors~main.chunk.js:22873:7)
    at runEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23545:7)
    at digestEffect (http://localhost:3000/static/js/vendors~main.chunk.js:23612:5)
    at next (http://localhost:3000/static/js/vendors~main.chunk.js:23502:9)
    at currCb (http://localhost:3000/static/js/vendors~main.chunk.js:23592:7)
End of invocation stack
    at NodeInternal.callUnary (http://localhost:3000/static/js/main.chunk.js:519:23)
    at async NodeInternal.stakingAccount (http://localhost:3000/static/js/main.chunk.js:889:12)
```

the main stack at the bottom is not any better. I'll remove the comment about switching to async+await

**Review by kostko — APPROVED** [review]

File: client-sdk/ts-web/core/src/client.ts [hunk 2]
```diff
@@ -585,18 +594,31 @@ export class GRPCWrapper {
...
+            // from oasis-core as expected above, try using JSON to stringify it so that we don't
+            // end up with [object Object].
+            const innerMessage = e instanceof Error ? e.toString() : JSON.stringify(e);
+            const message = `callUnary method ${method}: ${innerMessage}
+Invocation stack:
+${invocationStack}
+End of invocation stack`;
+            // @ts-expect-error options and cause not modeled
+            const wrapped = new Error(message, {cause: e});
+            throw wrapped;
```
Comment (lukaw3d) [inline client-sdk/ts-web/core/src/client.ts:621]: Output in Chrome:

error.toString()

- after:
  ```
  Error: callUnary method https://grpc.oasis.dev/oasis-core.Staking/Account: RpcError: Http response at 400 or 500 level
  Invocation stack
  at NodeInternal.callUnary (http://localhost:3000/static/js/vendors~main.chunk.js:3840:29)
  at NodeInternal.stakingAccount (http://localhost:3000/static/js/vendors~main.chunk.js:4242:17)
  ...40 lines
  End of invocation stack
  ```

- original error:
  ```
  RpcError: Http response at 400 or 500 level
  ```

error.stack

- after:
  ```
  Error: callUnary method https://grpc.oasis.dev/oasis-core.Staking/Account: RpcError: Http response at 400 or 500 level
  Invocation stack
  at NodeInternal.callUnary (http://localhost:3000/static/js/vendors~main.chunk.js:3840:29)
  at NodeInternal.stakingAccount (http://localhost:3000/static/js/vendors~main.chunk.js:4242:17)
  ...40 lines
  End of invocation stack
    at http://localhost:3000/static/js/vendors~main.chunk.js:3874:23
  ```

- original error:
  ```
  Error: Http response at 400 or 500 level
    at new E (index.js:21)
    at Y (index.js:56)
    at oc.<anonymous> (index.js:55)
    at Ib (index.js:33)
    at O (index.js:32)
    at wc (index.js:43)
    at yc (index.js:46)
    at oc.push../node_modules/grpc-web/index.js.n.W (index.js:44)
    at oc.push../node_modules/grpc-web/index.js.n.R (index.js:44)
  ```

error.cause?.stack

- after
  ```
  Error: Http response at 400 or 500 level
    at new E (index.js:21)
    at Y (index.js:56)
    at oc.<anonymous> (index.js:55)
    at Ib (index.js:33)
    at O (index.js:32)
    at wc (index.js:43)
    at yc (index.js:46)
    at oc.push../node_modules/grpc-web/index.js.n.W (index.js:44)
    at oc.push../node_modules/grpc-web/index.js.n.R (index.js:44)
  ```

Comment (lukaw3d) [reply client-sdk/ts-web/core/src/client.ts:621]: I don't like how much different this behaves than normal errors. I'd prefer
```ts
const wrapped = new Error(`callUnary method ${method}: ${e.message}`, { cause: e });
wrapped.stack += `\nInvocation stack:\n${invocationStack}\nEnd of invocation stack`;
throw wrapped;
```

wrapped.toString()
```
Error: callUnary method https://grpc.oasis.dev/oasis-core.Staking/Account: RpcError: Http response at 400 or 500 level
```

wrapped.stack
```
Error: callUnary method https://grpc.oasis.dev/oasis-core.Staking/Account: RpcError: Http response at 400 or 500 level
    at http://localhost:3000/static/js/vendors~main.chunk.js:3868:23
Invocation stack:
Error
    at NodeInternal.callUnary (http://localhost:3000/static/js/vendors~main.chunk.js:3840:29)
    at NodeInternal.stakingAccount (http://localhost:3000/static/js/vendors~main.chunk.js:4237:17)
    ...40 lines
End of invocation stack
```

Comment (pro-wh) [reply client-sdk/ts-web/core/src/client.ts:621]: I saw that .stack is non-standard. Reading it and potentially getting undefined is one thing. But should we really try setting it?

Comment (lukaw3d) [reply client-sdk/ts-web/core/src/client.ts:621]: or
```ts
const wrapped = new Error(`callUnary method ${method}: ${e.message}`, { cause: e });
wrapped.stack += `\nCause stack:\n${cause.stack}\nEnd of cause stack`;
wrapped.stack += `\nInvocation stack:\n${invocationStack}\nEnd of invocation stack`;
throw wrapped;
```

Comment (lukaw3d) [reply client-sdk/ts-web/core/src/client.ts:621]: :shrug:

Comment (lukaw3d) [reply client-sdk/ts-web/core/src/client.ts:621]: https://github.com/lukaw3d/run-in-everything/tree/f37fdfddbf9c4d6573cbcc1766fd54c43d046ad6 even webkit doesn't break

Comment (pro-wh) [reply client-sdk/ts-web/core/src/client.ts:621]: ok let's go with that

**Pull request merged**